core/src/autogluon/core/metrics/__init__.py

```diff
@@ -303,33 +362,69 @@ def rmse_func(predictions, targets):
 recall = make_scorer('recall',
                      sklearn.metrics.recall_score)
 
+
+def customized_log_loss(y_true, y_pred):
```

Comment:
Could you describe the need for this? How does this impact Tabular overall if the eval_metric is log_loss?

Reply:
Basically, it solves an issue in sklearn log_loss. The following script will raise an error for sklearn log_loss:

```python
from sklearn.metrics import log_loss

log_loss([0, 1, 0], [[0.1, 0.2, 0.7], [0.2, 0.7, 0.1], [0.7, 0.1, 0.2]])

```

Error:

```
-> 2232                                                   lb.classes_))
   2233         else:
   2234             raise ValueError('The number of classes in labels is different '

ValueError: y_true and y_pred contain different number of classes 2, 3. Please provide the true labels explicitly through the labels argument. Classes found in y_true: [0 1]
```

The root cause is that sklearn will try to infer the label set by looking at the ground truth labels. This is not needed because we have already inferred the label set in AutoGluon.

Reply:
Found a case where this still errors:

```
from autogluon.tabular import TabularPrediction as task

LABEL = 'class'
SAMPLE = 1000


if __name__ == '__main__':
    path_prefix = 'https://autogluon.s3.amazonaws.com/datasets/AdultIncomeBinaryClassification/'
    path_train = path_prefix + 'train_data.csv'
    path_test = path_prefix + 'test_data.csv'

    train_data = task.Dataset(file_path=path_train)
    test_data = task.Dataset(file_path=path_test)

    import pandas as pd
    with pd.option_context('display.max_columns', None, 'display.width', 1000):
        print(train_data)

    if SAMPLE is not None:
        train_data = train_data.sample(n=SAMPLE, random_state=0)

    predictor = task.fit(train_data=train_data, label=LABEL, hyperparameters='toy', eval_metric='log_loss')

    leaderboard = predictor.leaderboard(test_data, extra_info=True)

    test_data_only_pos = test_data[test_data[LABEL] == ' <=50K']

    # FIXME: This crashes
    predictor.leaderboard(test_data_only_pos)
```

Reply:
Nice catch. I previously only fixed the multi-class case. It seems that the binary case also contains this issue and I just submitted the fix.

Reply:
This code change appears to fix the issue:
```
def customized_log_loss(y_true, y_pred):
    """

    Parameters
    ----------
    y_true : array-like or label indicator matrix
        Ground truth (correct) labels for n_samples samples.

    y_pred : array-like of float
        The predictions. shape = (n_samples, n_classes) or (n_samples,)

    Returns
    -------
    loss
        The negative log-likelihood
    """
    assert y_true.ndim == 1
    if y_pred.ndim == 1:
        labels = [0, 1]
    else:
        assert y_pred.ndim == 2, 'Only ndim=2 is supported'
        labels = np.arange(y_pred.shape[1], dtype=np.int32)
    return sklearn.metrics.log_loss(y_true.astype(np.int32), y_pred, labels=labels)
```

Is there a reason we are converting to int32?

Reply:
I think both implementations are fine. For converting to int32, sometimes y_true can be float so I made the conversion.